### PR TITLE
Add subnet methods to the network domain service layer

### PR DESCRIPTION
**Pull request by nvinuesa**
This patch adds the rest of the subnet methods to the service layer on the new network domain.

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps
This is not wired-up yet so unit tests are enough:
```
 go test github.com/juju/juju/domain/network/... -gocheck.v
```

## Links

**Jira card:** JUJU-5105

### CODE REVIEW

File: domain/network/service/interface.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	// for moving subnets.
+	GetSubnetsByCIDR(ctx context.Context, cidrs ...string) (network.SubnetInfos, error)
+	// UpdateSubnet updates the subnet identified by the passed uuid.
+	UpdateSubnet(ctx context.Context, uuid string, spaceID string) error
+	// DeleteSubnet deletes the subnet identified by the passed uuid.
+	DeleteSubnet(ctx context.Context, uuid string) error
+	// UpsertSubnets updates or adds each one of the provided subnets in one
+	// transaction.
+	UpsertSubnets(ctx context.Context, subnets []network.SubnetInfo) error
+}
```
Comment (SimonRichardson) [inline domain/network/service/interface.go:49]: Can we create a `SpaceState` and a `SubneState` and then `State` is just a composition of those two?

Comment (nvinuesa) [reply domain/network/service/interface.go]: yeah I like it! see fd1e0f0b87

File: domain/network/service/interface.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+// State describes retrieval and persistence methods needed for the network
+// domain service.
+type State interface {
+	SpaceState
+	SubnetState
+}
+
+// SpaceState describes persistence layer methods for the space (sub-) domain.
+type SpaceState interface {
+	// AddSpace creates and returns a new space.
```
Comment (manadart) [inline domain/network/service/interface.go:21]: Except it only returns an error.

File: domain/network/service/subnet.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+		return "", errors.Trace(err)
+	}
+
+	return network.Id(uuid.String()), nil
+}
+
+// Subnet returns the subnet identified by the input UUID,
+// or an error if it is not found.
+func (s *SubnetService) Subnet(ctx context.Context, uuid string) (*network.SubnetInfo, error) {
+	return s.st.GetSubnet(ctx, uuid)
```
Comment (manadart) [inline domain/network/service/subnet.go:44]: We should be tracing/annotating the errors for these, so at least
```go
info, err := s.st.GetSubnet(ctx, uuid)
return info, errors.Trace(err)
```

File: domain/network/state/space_test.go [hunk 4]
```diff
@@ -33,8 +33,8 @@ func (s *stateSuite) TestAddSpace(c *gc.C) {
 	c.Assert(err, jc.ErrorIsNil)
 	err = st.AddSubnet(
 		ctx.Background(),
+		subnetUUID.String(),
 		network.SubnetInfo{
-			ID:                network.Id(subnetUUID.String()),
```
Comment (manadart) [inline domain/network/state/space_test.go:37]: I actually like the idea of upstream caller being able to supply this ID, as that's what I want to happen eventually.

We can land this as is, but I'd like a change for this and other services to:
- Just work if the incoming info has a UUID.
- Populate it with a generated one if empty.

The caller always gets the UUID inserted, so this is safe.

**Review by manadart — COMMENTED** [review]
Some suggestions.

File: domain/network/state/space_test.go (hunk 4 above)
Comment (nvinuesa) [reply domain/network/state/space_test.go]: Yeah indeed I think it's a good convention so I did it for this service (only AddSubnet has this type of input). See 8c41e31aac

File: domain/network/service/subnet.go (hunk 3 above)
Comment (nvinuesa) [reply domain/network/service/subnet.go]: done, see 8c41e31aac

**Review by manadart — APPROVED** [review]
Thanks for the changes.

**Pull request merged**